zug/util.hpp

```diff
@@ -91,7 +92,8 @@ constexpr decltype(auto) invoke_composition(Fns&& fns, Args&&... args)
 template <typename Fn, typename... Fns>
 struct composed
 {
-    std::tuple<Fn, Fns...> fns;
+    using function_tuple = std::tuple<Fn, Fns...>;
+    function_tuple fns;
```

Comment:
Idea, inherit from the tuple instead of having it as member. Allows for empty `composed` objects.

Reply:
If it's OK with you, I'd suggest we'd look at that when addressing the `operator|` functionality, which can include the suggestion for the empty `composed` you made. I'll keep this in mind for then, and address the other blockers for that (e.g., the template args `<Fn, Fn...>`).

Reply:
The empty based is unrelated to the other suggestion. Because if `Fns` are empty themselves, `std::tuple` is empty (uses empty_base optiization). Anyways yeah we can do that in a separate step if you prefer. It was just a suggestion. (I mean here by empty I mean with size 0, not w/o functions)

Reply:
Ah I see what you mean now. OK, I can do that here.

Reply:
Ah I see - then I misunderstood. I wasn’t aware of this optimization - thanks for pointing it out! I will address this here.